imx_pwm_led.c

```diff
@@ -0,0 +1,2069 @@
+
+	cr = PWMCR_REPEAT(repeat) | PWMCR_DOZEEN | PWMCR_WAITEN | PWMCR_DBGEN |
+	     PWMCR_CLKSRC_IPG_HIGH | PWMCR_PRESCALER(prescaler) |
+	     PWMCR_POUTC(invert);
+
+	/* Software reset */
+	__raw_writel(PWMCR_SWR, self->mmio_base + PWMCR);
+	udelay(10);
+	/* Activate & de-activate PWM (seems to be necessary after a reset) */
+	__raw_writel(PWMCR_EN, self->mmio_base + PWMCR);
+	__raw_writel(0, self->mmio_base + PWMCR);
+
+	/* Silence */
+	__raw_writel(0, self->mmio_base + PWMSAR);
+
+	clk = clk_get_rate(self->clk_ipg);
+	/* Sanity check of IPG clock frequency: */
+	if (clk < 1000000) {
+		dev_err(self->dev, "%s: invalid IPG clock: %u", __func__, clk);
+		return -EFAULT;
+	}
+	/* Warn if the PWM FIFO will either run full or empty:
+	 * - If it runs full, samples will be lost, possibly including the last
+	 *   one, which will leave the duty cycle at an unexpected value.
+	 * - If it runs empty, a hardware bug/feature of the i.MX can cause
+	 *   glitches in the output. While there is no official erratum, this
+	 *   issue is discussed here: https://community.nxp.com/thread/356855
+	 *   For unaffected devices, this warning can be safely ignored. */
+	if (((uint64_t)sample_rate * (uint64_t)period *
+		     (uint64_t)(prescaler + 1)
+	     << repeat) != clk) {
+		dev_warn(self->dev,
+			 "%s: sample_rate %u Hz * period %u * "
+			 "(prescaler %u + 1) << repeat %u != ipg_clk %u Hz",
+			 __func__, sample_rate, period, prescaler, repeat, clk);
+	}
+	self->pwm_fifo_period_us =
+		(period * (prescaler + 1) << repeat) / (clk / 1000000);
+	dev_dbg(self->dev, "%s: setting PWMPR: 0x%08x (%u.%03u Hz)", __func__,
+		period - 2, clk / (period * (prescaler + 1)),
+		((clk % (period * (prescaler + 1))) * 1000) / period);
+	__raw_writel(period - 2, self->mmio_base + PWMPR);
+
+	/* Configure */
+	dev_dbg(self->dev, "%s: setting PWMCR: 0x%08x", __func__, cr);
+	__raw_writel(cr, self->mmio_base + PWMCR);
+	return 0;
+}
+
+static inline void enable_pwm(struct imx_pwm_led *self)
+{
+	u32 cr = __raw_readl(self->mmio_base + PWMCR);
+	cr |= PWMCR_EN;
+	__raw_writel(cr, self->mmio_base + PWMCR);
+}
+
+static void disable_pwm(struct imx_pwm_led *self)
+{
+	u32 cr = 0;
+	dev_dbg(self->dev, "%s", __func__);
+	/* Disable PWM */
+	__raw_writel(cr, self->mmio_base + PWMCR);
+	/* Release clocks */
+	clk_disable_unprepare(self->clk_per);
+	clk_disable_unprepare(self->clk_ipg);
+}
+
+static inline void set_pwm_duty(struct imx_pwm_led *self, uint duty)
+{
+	__raw_writel(duty, self->mmio_base + PWMSAR);
+}
+
+static int wait_pwm_fifo_empty(struct imx_pwm_led *self)
+{
+	uint i;
+	uint level;
+	/* Make two attempts at waiting for the FIFO to empty: */
+	for (i = 0; i < 2; i++) {
+		uint min_sleep;
+		/* Get the number of samples in the PWM FIFO: */
+		level = __raw_readl(self->mmio_base + PWMSR) &
+			PWMSR_FIFOAV_MASK;
+		/* If the FIFO is empty:*/
+		if (level == 0) {
+			return 0;
+		}
+		/* Sleep until the FIFO should be empty: */
+		min_sleep = level * self->pwm_fifo_period_us;
```

Comment:
Should probably have an upper bound for the sleep, you wouldn't want to sleep for more than 100ms.

Reply:
Fixed